### PR TITLE
BUG: make order of index from pd.concat deterministic

### LINKED ISSUE #17344: Unconsistent (random) behaviour of pd.concat with different indexes
#### Code Sample, a copy-pastable example if possible

The following code

```python
import pandas as pd

dfs_sq = []

dfs_sq.append(pd.DataFrame(index=[0, 'sess'], columns=range(1,3)))
for i in range(5):
    dfs_sq.append(pd.DataFrame(index=[0, 1, 'sess'], columns=range(1,3)))

df_sq = pd.concat(dfs_sq, axis=1)

assert df_sq.index[1] == 'sess', df_sq.index
```
... saved as ``dependable.py``, succeeds if called as ``PYTHONHASHSEED=40 python3 dependable.py`` and fails if called as ``PYTHONHASHSEED=41 python3 dependable.py``:
```bash
Traceback (most recent call last):
  File "dependable.py", line 11, in <module>
    assert df_sq.index[1] == 'sess', df_sq.index
AssertionError: Index([0, 1, 'sess'], dtype='object')
```
#### Problem description

This is not very nice.

#### Expected Output

The same - which one is not particularly important.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.5.3.final.0
python-bits: 64
OS: Linux
OS-release: 4.9.0-3-amd64
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: it_IT.UTF-8
LOCALE: it_IT.UTF-8

pandas: 0.21.0.dev+389.g276f3089a
pytest: 3.0.6
pip: 9.0.1
setuptools: None
Cython: 0.25.2
numpy: 1.12.1
scipy: 0.19.0
pyarrow: None
xarray: None
IPython: 5.1.0.dev
sphinx: 1.5.6
patsy: 0.4.1
dateutil: 2.6.0
pytz: 2017.2
blosc: None
bottleneck: 1.2.1
tables: 3.3.0
numexpr: 2.6.1
feather: 0.3.1
matplotlib: 2.0.2
openpyxl: None
xlrd: 1.0.0
xlwt: 1.1.2
xlsxwriter: 0.9.6
lxml: None
bs4: 4.5.3
html5lib: 0.999999999
sqlalchemy: 1.0.15
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: 0.2.1

</details>

### PR DESCRIPTION
**Pull request by toobaz**
- [x] closes #17344
- [x] tests added / passed
- [x] passes `git diff master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v0.21.0.txt [hunk 1]
```diff
@@ -405,6 +405,7 @@ Reshaping
 - Bug in :func:`crosstab` where passing two ``Series`` with the same name raised a ``KeyError`` (:issue:`13279`)
 - :func:`Series.argmin`, :func:`Series.argmax`, and their counterparts on ``DataFrame`` and groupby objects work correctly with floating point data that contains infinite values (:issue:`13595`).
 - Bug in :func:`unique` where checking a tuple of strings raised a ``TypeError`` (:issue:`17108`)
+- Bug in :func:`concat` which would randomly determine the order of the index along the common dimension (:issue:`17344`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: this is not very imformative as this is a very specific case that you are fixing ( it makes it seem like this is a very common / huge problem)

**Review by jreback — APPROVED** [review]
just revise the whatsnew, otherwise lgtm. ping on green.

**Comment by toobaz** [comment]
@jreback ping

**Comment by jreback** [comment]
thanks!

**Pull request merged**